# OBJECTIVE

Assume you run a small security company that provides physical security services in the area and you recently won a new contract in the area to provide 24x7 security to a small building under construction.  For simplicity we will design the solution for only 24 hours, but if you want to go above and beyond, feel free to write code that handles the 24x7 scenario as well.

You have 6 security guards available at the moment who you can assign to this building but your goal is to make more money out of this contract and spend less in wages (hence greedy!)

The cost/wage structure is as follows:
People working less than or equal to 8 hours will be paid 15 dollars/hr.
Anyone working overtime (>8 hours) will be paid an additional 5 dollars, i.e. 20 dollars/hr.

# Define a Function

In [1]:
# Create a greedy algorithm. 
def greedy(time_needed=24*7, shift=8):
    
    # First input parameter: Contract time needed. Default is 24x7
    # Second input paramter: One work shift is set to 8 hours to avoid any overtime cost.
    
    # When the shift is less than or equal to 8 hours: 
    if shift >0 and shift <=8: 
        security_list = ['A', 'B', 'C', 'D', 'E', 'F']  # Set a list of securities. 
        total_cost = 0  # Initial total cost. 
        shift = 8  # One work shift is set to 8 hours to avoid any overtime cost. 
        cost_hr = 15  # Hourly cost is $15. 
        security_needed = []  # Total securities needed, which can be repeated if needed. 
        security_hours = {}  # Total working hours for each security, with security as key and hours as value.  


        while time_needed > 0:  
            for security in security_list: # Loop the security list. Start from the first security. 
                if time_needed > shift:  
                    time_needed -= shift  
                    wage = shift * cost_hr
                    total_cost += wage
                    security_needed.append(security)
                    security_hours[security] = security_hours.get(security, 0) + shift
                else: 
                    total_cost = total_cost + (time_needed * cost_hr)
                    security_needed.append(security)
                    security_hours[security] = security_hours.get(security, 0) + time_needed
                    time_needed = 0
                    break

        print(f'Total cost is: ${total_cost}.')
        print(f'\nWork shift: {shift} hours.')
        print(f'\nTotal securities needed: {len(security_needed)}.')
        print(f'\nSecurity Schedule:{security_needed}.')
        print(f'\nSecurity Hours:{security_hours}.')
    
    # When the shift is outside of (0,8) (not including 8): 
    else: 
        print("Warning: Shift should be equal to or less than 8 hours to save cost!")
    
greedy()

Total cost is: $2520.

Work shift: 8 hours.

Total securities needed: 21.

Security Schedule:['A', 'B', 'C', 'D', 'E', 'F', 'A', 'B', 'C', 'D', 'E', 'F', 'A', 'B', 'C', 'D', 'E', 'F', 'A', 'B', 'C'].

Security Hours:{'A': 32, 'B': 32, 'C': 32, 'D': 24, 'E': 24, 'F': 24}.


# CONCLUSION

- Only work shift less than the overtime threshold is considered in my greedy algorithm. In this case, we assign the work shift to 8 hours. Note that the work shift can be less than 8 hours if needed. It is greedy because we are trying to avoid any overtime hours to minimize the wage costs. There are 6 securities available, expressed as a list from A to F. The algorithm will loop over these 6 securities starting from A to F. If there are more shifts are needed after going through the security list, we will start again from A, until there is no more hour needed. Note that this offers the most potential working opportunities to A, following by B, C, D, etc. We can be greedy to put the best performance security as A, while the worest performance as F. All picked securities except the last picked will get an 8-hour shift. At any contract hour, we are greedy to make sure that there will always be only one security at shift. The cost calculation is the total hours needed times the cost per hour. Since overtime is never allowed in this algorithm, the hourly rate is consistant at 15 dollars. 

- This greedy algorithm contains a nested loop: a while loop and a for loop. Thus, the time complexity(big O notation) is O(n*n). 

- The greedy algorithm provides the most cost effeceive solution to avoid any overtime or overlapped schedules (avoid more than one security working at a time). There are other alternative solutions to this problem, split the same amount of working hours to all securities for any given contract hour while avoiding any overtime hours for each shift, for example. However, I think my algorithm works the best of them, because it ensures every picked security gets 8 hours shift when they go to work (except the last picked security), while the security company do not need to pay any overtime costs. Also, since the security with the best performance can get more potential work opportunities, it encourages good competition withtin the security team. 

- Even if the inputs are different in this scenario, my algorithm would still work. The first default parameter shows 24x7 contract hours and the second defulat parameter is 8 hours per shift, with an output of 2,520 dollars of total cost and 21 securities(can be repeatedly looping the security list). Even though when we change the contract hours to 25, this algorithm will still work, except that the last picked security would only get a 1 hour shift. Any shorter shifts than 8 hours would work as clean. Contract times and work shift are parameters in the algorithm. As long as the work shift does not meet the overtime hour threshold (8 hours), this algorithm would work. Otherwise it would print a warning message. 

- If I were not constrained to a greedy algorithm, a potential approach is to use an equal system to give all securities the same working hours for any given contract hours. Another potential approach is to use set/subsets. 

Generally, data engineers can easily utilize greedy algorithm for any given np problem. In addition, it is easier to analyze runtimes for greedy algorithm than other techniques. Also, greedy algorithm can output an approximate answer easily instead of taking other complixated approaches. 